feat: Migrate Pipedrive CRM from Revert to native Pipedrive APIs

itsalam opened the pull request:
- Changed OAuth flow from environment varibles to user supplied on installation.
- Added apis to reflect User OAuth redirects
- Added setup page with instrutions
- Added dayjs and pipedrive package to the pipedrive package

## What does this PR do?

- Fixes #16797 (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description) 
  - No Linear Issue number submitted; N/A.

Loom Videos:
Adding the app: https://www.loom.com/share/d3cf7e794366418b945aacb6e599463e?sid=d926b675-02de-4d49-9aa7-052c49ddcac9
Editing events: https://www.loom.com/share/1358d47527cf4b6bb922bdd4cf42c7be?sid=21842087-03d9-43ab-85a0-2e525d27b3cf
<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.
  - Seems difficult to test with current available methods. 

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
  - None
- What are the minimal test data to have?
  - ngrok'd environment, Pipedrive test app/account 
- What is expected (happy path) to have (input and output)?
  -  Event should be created and reflect changes on Pipedrive.
- Any other important info that could help to test that PR

itsalam committed:
- Changed Pipedrive CRMService Apis from Revert to Pipedrives

- Changed OAuth flow from environment varibles to user supplied on installation.

- Added apis to reflect User OAuth redirects

- Added setup page with instrutions

- Added dayjs and pipedrive package to the pipedrive package

itsalam committed:
fixed redirect on install

Praashh commented on packages/app-store/pipedrive-crm/lib/util.ts:
```diff
@@ -0,0 +1,70 @@
+  const oAuthConfig = new OAuth2Configuration({
+    clientId: keys.client_id,
+    clientSecret: keys.secret_key,
+    redirectUri: "https://9ba4a35f0b7b.ngrok.app/api/integrations/pipedrive-crm/callback", //`${WEBAPP_URL}/api/integrations/pipedrive-crm/callback`,
```
redirect URL shouldn't be hard coded.

itsalam replied:
Apologies, just needed to hardcode it to properly test it. Fixing now

itsalam committed:
Fixed hardcoded URL and updated Readme

Praashh commented on packages/app-store/_pages/setup/index.tsx:
```diff
@@ -16,6 +16,7 @@ export const AppSetupMap = {
   sendgrid: dynamic(() => import("../../sendgrid/pages/setup")),
   stripe: dynamic(() => import("../../stripepayment/pages/setup")),
   paypal: dynamic(() => import("../../paypal/pages/setup")),
+  "pipedrive-crm": dynamic(() => import("../../pipedrive-crm/pages/setup")),
```
let's keep this key same as other keys.
```suggestion
  pipedrive-crm: dynamic(() => import("../../pipedrive-crm/pages/setup")),
```

itsalam replied:
This change needs quotes becase of the dash "-" in its key, as noted by the other keys with dashes as well.

Praashh replied:
yea got it

Praashh commented on packages/app-store/_pages/setup/index.tsx:
```diff
@@ -16,6 +16,7 @@ export const AppSetupMap = {
   sendgrid: dynamic(() => import("../../sendgrid/pages/setup")),
   stripe: dynamic(() => import("../../stripepayment/pages/setup")),
   paypal: dynamic(() => import("../../paypal/pages/setup")),
+  "pipedrive-crm": dynamic(() => import("../../pipedrive-crm/pages/setup")),
```
let's keep this key same as other keys.
```suggestion
  pipedrive-crm: dynamic(() => import("../../pipedrive-crm/pages/setup")),
```

Praashh commented on packages/app-store/pipedrive-crm/api/add.ts:
```diff
@@ -1,37 +1,50 @@
   const { teamId } = req.query;
   const user = req.session?.user;
-  if (!user) {
+  if (user === undefined) {
```
```suggestion
  if (!user) {
```

much readable.

itsalam replied:
much ack'd

Praashh commented on packages/app-store/pipedrive-crm/api/callback.ts:
```diff
@@ -1,19 +1,54 @@
+
+  const credKeys = pipedriveAppKeysSchema.parse(credentials.key);
+
+  await prisma.credential.update({
+    where: {
+      id: credentials.id,
+    },
+    data: {
+      key: {
+        ...credKeys,
+        tokens,
+        lastUpdated: new Date(),
+      },
+    },
+  });
```
there should be checking for updating the credential, what if it fails?

itsalam replied:
good point, ill update to catch that

Praashh commented on packages/app-store/pipedrive-crm/api/setup.ts:
```diff
@@ -0,0 +1,22 @@
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  if (req.method !== "POST") return res.status(405).json({ message: "Method not allowed" });
+  // Check that user is authenticated
```
don't think we need this comment here.

itsalam committed:
Addressed Fixes
- Removed unecessary comments
- Added better error handling in the callback
- More readble if condition in pipedrive/api/add

Praashh reviewed (changes requested):
we also need updated steps for pipedrive in `packages/app-store/pipedrive-crm/README.md` and `.env.appStore.example`. please add

itsalam committed:
removed .env references for revert

itsalam committed:
updated readme and turbo.json

Praashh commented:
Hey @itsalam, thank you so much for your contribution but for now, we are going with the other one.

rhinck commented:
> Hey @itsalam, thank you so much for your contribution but for now, we are going with the other one.

Hi @Praashh, is hosted cal.com using a Pipedrive integration without bugs currently? I see issues when trying to link our org:

![image](https://github.com/user-attachments/assets/a9f7f65c-e755-4cbc-9018-282a7a36ca5f)
